### PR TITLE
Update to JUnit 5.9 (#652 / 657)

### PR DESCRIPTION
**Pull request by Bukama**
Proposed commit message:

```
Update to JUnit Jupiter 5.9 (#652 / #657)

Updates the used JUnit Jupiter version

closes: #652
PR: #657
```

---
**PR checklist**

The following checklist shall help the PR's author, the reviewers and maintainers to ensure the quality of this project.
It is based on our contributors guidelines, especially the ["writing code" section](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#writing-code).
It shall help to check for completion of the listed points.
If a point does not apply to the given PR's changes, the corresponding entry can be simply marked as done.

### CODE REVIEW

File: gradle.properties
Comment (beatngu13) [inline gradle.properties:1]: As you know, this actually bumps the JUnit version we use under hood. It not just incorporates it in our build (as suggested in #652). From our [contributing guidelines](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#updating-junit-5):

>As [documented](https://github.com/junit-pioneer/junit-pioneer/blob/main/README.md#dependencies) Pioneer aims to use the lowest JUnit 5 version that supports Pioneer's feature set. At the same time, there is no general reason to hesitate with updating the dependency if a new feature requires a newer version or the old version has a severe bug.

Our code currently only fails with JUnit 5.9 due to test internals, not because we need a certain feature or fix. I'm cool with it, I just wonder if we can and/or want to stay on 5.8.2?

In any case, please also update the build accordingly:

* https://github.com/junit-pioneer/junit-pioneer/blob/main/.github/workflows/build.yml#L121
* https://github.com/junit-pioneer/junit-pioneer/blob/main/.github/workflows/build.yml#L150

**Review by beatngu13 — CHANGES_REQUESTED** [review]

File: gradle.properties
Comment (Bukama) [reply gradle.properties:1]: You are totally right (e.g. #320 was the issue about upgrading to 5.6.0). But I understood your comment in the issue (#652) as a "we need it" and therefore created this PR. I'll check if I can get the build work without updating. If not I would call a vote on next maintainer meeting about either upgrading the Jupiter version or not adding it to the build matrix to avoid failing builds.

P.S. Will update the two sections of the workflow script you've linked while doing this.

Comment (Bukama) [reply gradle.properties:1]: Okay in 5.9.0 the interface returns `org.junit.jupiter.api.extension.ExecutableInvoker`, but in 5.8.2 the class is placed `org.junit.jupiter.engine.execution.ExecutableInvoker`. So even if I "overwrite" the `getExecutableInvoker` Methode it would return the false interface.

File: src/test/java/org/junitpioneer/internal/TestExtensionContext.java
Comment (Michael1993) [inline src/test/java/org/junitpioneer/internal/TestExtensionContext.java:51]:
```suggestion
	// @Override once we baseline against 5.9
```
We don't have to update, so we should not.

File: .github/workflows/build.yml
Comment (beatngu13) [inline .github/workflows/build.yml:121]: You can / have to remove 5.8.2, as our test code is now incompatible.

```suggestion
        junit-version: [ '5.9.0' ]
```

File: .github/workflows/build.yml
Comment (beatngu13) [inline .github/workflows/build.yml:150]:
```suggestion
        junit-version: [ '5.9.0' ]
```

**Review by beatngu13 — CHANGES_REQUESTED** [review]

**Review by beatngu13 — APPROVED** [review]

**Comment by Bukama** [comment]
> Abstract entry-based extension > should not mix backups of different extensions on clear environment variable and set system property FAILED
    org.junit.jupiter.api.extension.ExtensionConfigurationException: Cannot access Java runtime internals to modify environment variables. Have a look at the documentation for possible solutions: https://junit-pioneer.org/docs/environment-variables/#warnings-for-reflective-access
        at app//org.junitpioneer.jupiter.EnvironmentVariableUtils.getFieldValue(EnvironmentVariableUtils.java:104)
        at app//org.junitpioneer.jupiter.EnvironmentVariableUtils.setInProcessEnvironmentClass(EnvironmentVariableUtils.java:78)
        at app//org.junitpioneer.jupiter.EnvironmentVariableUtils.modifyEnvironmentVariables(EnvironmentVariableUtils.java:50)
        at app//org.junitpioneer.jupiter.EnvironmentVariableUtils.clear(EnvironmentVariableUtils.java:45)
        at app//org.junitpioneer.jupiter.AbstractEntryBasedExtensionTests.setUp(AbstractEntryBasedExtensionTests.java:36)

        Caused by:
        java.lang.reflect.InaccessibleObjectException: Unable to make field private static final java.util.HashMap java.lang.ProcessEnvironment.theEnvironment accessible: module java.base does not "opens java.lang" to unnamed module @59d016c9
            at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
            at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
            at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:180)
            at java.base/java.lang.reflect.Field.setAccessible(Field.java:174)
            at org.junitpioneer.jupiter.EnvironmentVariableUtils.getFieldValue(EnvironmentVariableU

**Comment by Bukama** [comment]
@beatngu13  Do you have an idea how to implement the set environment thing without reflection so I can get rid of the error?

**Comment by beatngu13** [comment]
These tests are already failing on `main`:

https://github.com/junit-pioneer/junit-pioneer/runs/7782024824?check_suite_focus=true#step:8:61

So either you ignore it, as you didn't change anything in this regard, or you disable the test on Java 17+ (preferred by me):

```java
@EnabledForJreRange(max = JRE.JAVA_16, disabledReason = "See: https://github.com/junit-pioneer/junit-pioneer/issues/509")
```

In #651 / #614, we try to (re)enable them with `--adds-open`, but we are still struggling with the Gradle setup.

**Comment by Bukama** [comment]
Disabled `AbstractEntryBasedExtensionTests`. The two `EnvironmentVariable...Tests` were already disabled (what's strange as the `AbstractEntryBasedExtensionTests` also belongs to these env thing. Let's see.

**Comment by Bukama** [comment]
Blocked by #651 (Gradle 7.5) due Java 18 is not supported with 7.4:
Gradle-issue https://github.com/gradle/gradle/issues/20344 is fixed with 7.5

**Comment by beatngu13** [comment]
@Bukama not blocked, all builds passed (now). Feel free to merge. ✌️

**Pull request merged**
